pkg/query-service/app/http_handler.go

```diff
@@ -2572,19 +2620,83 @@ func (aH *APIHandler) execPromQueries(ctx context.Context, metricsQueryRangePara
 	return res, nil, nil
 }
 
-func (aH *APIHandler) queryRangeV3(queryRangeParams *v3.QueryRangeParamsV3, w http.ResponseWriter, r *http.Request) {
+func (aH *APIHandler) getLogFieldsV3(ctx context.Context, queryRangeParams *v3.QueryRangeParamsV3) (map[string]v3.AttributeKey, error) {
+	data := map[string]v3.AttributeKey{}
+	for _, v := range queryRangeParams.CompositeQuery.BuilderQueries {
+		if v.DataSource == v3.DataSourceLogs {
+			fields, apiError := aH.reader.GetLogFields(ctx)
+			if apiError != nil {
+				return nil, apiError.Err
+			}
+
+			// top level fields meta will always be present in the frontend. (can be support for that as enchancement)
+			getType := func(t string) (v3.AttributeKeyType, bool) {
+				if t == "attributes" {
+					return v3.AttributeKeyTypeTag, false
+				} else if t == "resources" {
+					return v3.AttributeKeyTypeResource, false
+				}
+				return "", true
+			}
+
+			for _, v := range fields.Selected {
```

Comment:
There is a `v` here and in the outer loop, making it hard to follow. Please use distinct/descriptive names to make it easier for the reader.